argoproj-labs/argocd-operator
# feat: unify SSO configuration under `.spec.sso` in backward-compatible way

**What type of PR is this?**
> Uncomment only one ` /kind` line, and delete the rest.
> For example, `> /kind bug` would simply become: `/kind bug`

> /kind bug
> /kind chore
> /kind cleanup
> /kind failing-test
 /kind enhancement
> /kind documentation
> /kind code-refactoring


**What does this PR do / why we need it**:
This PR re-introduces the effort to unify existing SSO configuration options for dex and keycloak under `.spec.sso`. It intoduces 
`.spec.sso.dex` and `.spec.sso.keycloak`. It also maintains backward compatibility to support `.spec.dex`, `DISABLE_DEX` and existing `.spec.sso` fields for keycloak configuration until they are deprecated in v0.6.0. It emits events warning users of these soon-to-be-deprecated config options.
Since we need to support both the old and new config specs for dex and keycloak, this PR introduces a wide range of checks in order to ensure there is no mismatch or illegal combinations of the various SSO spec options available 

**IMPORTANT NOTES**
1. SInce https://github.com/argoproj-labs/argocd-operator/pull/615 was merged, dex health check will fail if dex is running but is not configured. Therefore this PR introduces a new constraint where dex cannot be enabled (either through `.spec.ssp.provider=dex` or `DISABLE_DEX=false`) without providing some kind of configuration in `.spec.dex` or `.spec.sso.dex`
2. if using the env var, the absence of `DISABLE_DEX` no longer implies that dex is enabled by default. It will be treated as saying dex is disabled instead. In order not to break existing workloads, if dex pods are found to be running but there is no `DISBALE_DEX` flag set, the dex resources not be deleted if any dex configuration is found to exist (either `openShiftOAuth=true` or `dex.config` is supplied). In all other cases, dex resources will be deleted

**Have you updated the necessary documentation?**


**Which issue(s) this PR fixes**:
https://github.com/argoproj-labs/argocd-operator/issues/653

Fixes https://issues.redhat.com/browse/GITOPS-1332

**How to test changes / Special notes to the reviewer**:
1. launch operator locally with `make run`
3. Create empty argo-cd instance 
```cat <<EOF | kubectl apply -f - 
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: argocd
spec: {}
EOF
```
4. Since `DISABLE_DEX` was not set, check that dex resources don't come up by default
5. Set  `.spec.sso.provider=dex` 
6. Observe dex deployment doesn't come up. Observe error in logs indicating dex must also be configured by supplying a `.spec.sso.dex` 
7. Set `.spec.sso.dex.Config="test"`
8. Observe dex deployment comes up successfully 
9. Set `.spec.sso.image=test-image`
10. Observe error in logs indicating `.spec.sso` fields (for keycloak) cannot be specified when `.spec.sso.provider=dex`
11. Remove `.spec.sso.image=test-image`. Set `.spec.sso.keycloak.image=test-image`
12. Observe error in logs indicating `.spec.sso.keycloak` cannot be specified when `.spec.sso.provider=dex`
13. Remove .spec.sso.keycloak`. Set `.spec.dex.config=test-2`
14. Observe error in logs indicating `.spec.dex` fields cannot be specified when `.spec.sso.provider=dex`
15. Remove `.spec.dex`. Remove `.spec.sso.dex`. Set `.spec.sso.provider=keycloak`
16. Observe that dex resources are deleted and keycloak resources are created successfully
17. Set `.spec.sso.image=<some-keycloak-image>` and observe it gets updated
18. Set `.spec.sso.keycloak.image=xyz`. 
19. Observe error in logs indicating that conflicting information cannot be provided in `.spec.sso` fields and equivalent `.spec.sso.keycloak` fields 
20. Remove `.spec.sso` keycloak fields (`.spec.sso.image, .spec.sso.version, .spec.sso.resources, .spec.sso.verifyTLS`) and `.spec.sso.keycloak` 
21. Set `.spec.sso.dex.config=test`
22. Observe error in logs that `.spec.sso.dex` cannot be specified when `.spec.sso.provider=keycloak`
23. Remove `.spec.sso.dex`. Set `.spec.dex.openShiftOAuth=true`
24. Observe error in logs indicating that multiple SSO providers cannot be configured simultaneously 
25. Remove `.spec.sso` entirely and observe that keycloak resources are deleted 
=====================================================================================
1. launch operator locally with `make run DISABLE_DEX=false`
2. Create empty argo-cd instance 
```cat <<EOF | kubectl apply -f - 
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: argocd
spec: {}
EOF
```
27. Notice dex pod doesn't come up. Observe error in logs indicating that dex must be configured via `.spec.dex` if dex is enabled
28. Set `.spec.dex.openShiftOAuth=true`
29. Observe that dex pod comes up successfully
30. Set `.spec.dex.openShiftOAuth=false` and `.spec.sso.provider=keycloak` 
31. Observe keycloak pod should come up in order to preserve existing behavior and not introduce a breaking change
32. Remove `.spec.sso.provider=keycloak`, Set `.spec.sso.keycloak.image=test`
33. Observe error in logs indicating that SSO provider spec cannot be supplied when `.spec.sso.provider="`
34. Remove `.spec.sso.provider` and `.spec.sso.keycloak`
35. Set `.spec.sso.image=test`
36. Observe no errors in logs
========================================================================================
37. Execute `kubectl get events`
38. Observe that deprecation events were emitted warning users that `.spec.dex`, `DISABLE_DEX` and `.spec.sso` fields for key cloak have been deprecated and support will be removed in a future release
